Include/cpython/pyatomic.h

```diff
@@ -0,0 +1,397 @@
+// or relaxed atomics.
+
+// Atomically adds `value` to `address` and returns the previous value
+static inline int
+_Py_atomic_add_int(int *address, int value);
+
+static inline int8_t
+_Py_atomic_add_int8(int8_t *address, int8_t value);
+
+static inline int16_t
+_Py_atomic_add_int16(int16_t *address, int16_t value);
+
+static inline int32_t
+_Py_atomic_add_int32(int32_t *address, int32_t value);
+
+static inline int64_t
+_Py_atomic_add_int64(int64_t *address, int64_t value);
+
+static inline intptr_t
+_Py_atomic_add_intptr(intptr_t *address, intptr_t value);
+
+static inline unsigned int
+_Py_atomic_add_uint(unsigned int *address, unsigned int value);
+
+static inline uint8_t
+_Py_atomic_add_uint8(uint8_t *address, uint8_t value);
+
+static inline uint16_t
+_Py_atomic_add_uint16(uint16_t *address, uint16_t value);
+
+static inline uint32_t
+_Py_atomic_add_uint32(uint32_t *address, uint32_t value);
+
+static inline uint64_t
+_Py_atomic_add_uint64(uint64_t *address, uint64_t value);
+
+static inline uintptr_t
+_Py_atomic_add_uintptr(uintptr_t *address, uintptr_t value);
+
+static inline Py_ssize_t
+_Py_atomic_add_ssize(Py_ssize_t *address, Py_ssize_t value);
+
+// Performs an atomic compare-and-exchange. If `*address` and `expected` are equal,
+// then `value` is stored in `*address`. Returns 1 on success and 0 on failure.
+// These correspond to the "strong" variations of the C11 atomic_compare_exchange_* functions.
+static inline int
+_Py_atomic_compare_exchange_int(int *address, int expected, int value);
+
+static inline int
+_Py_atomic_compare_exchange_int8(int8_t *address, int8_t expected, int8_t value);
+
+static inline int
+_Py_atomic_compare_exchange_int16(int16_t *address, int16_t expected, int16_t value);
+
+static inline int
+_Py_atomic_compare_exchange_int32(int32_t *address, int32_t expected, int32_t value);
+
+static inline int
+_Py_atomic_compare_exchange_int64(int64_t *address, int64_t expected, int64_t value);
+
+static inline int
+_Py_atomic_compare_exchange_intptr(intptr_t *address, intptr_t expected, intptr_t value);
+
+static inline int
+_Py_atomic_compare_exchange_uint(unsigned int *address, unsigned int expected, unsigned int value);
+
+static inline int
+_Py_atomic_compare_exchange_uint8(uint8_t *address, uint8_t expected, uint8_t value);
+
+static inline int
+_Py_atomic_compare_exchange_uint16(uint16_t *address, uint16_t expected, uint16_t value);
+
+static inline int
+_Py_atomic_compare_exchange_uint32(uint32_t *address, uint32_t expected, uint32_t value);
+
+static inline int
+_Py_atomic_compare_exchange_uint64(uint64_t *address, uint64_t expected, uint64_t value);
+
+static inline int
+_Py_atomic_compare_exchange_uintptr(uintptr_t *address, uintptr_t expected, uintptr_t value);
+
+static inline int
+_Py_atomic_compare_exchange_ssize(Py_ssize_t *address, Py_ssize_t expected, Py_ssize_t value);
+
+static inline int
+_Py_atomic_compare_exchange_ptr(void *address, void *expected, void *value);
```

Comment:
Should it be `void** address` for clarity?

Reply:
The problem is that `void **` requires an explicit cast for almost every use, because things like `PyObject **` are not implicitly convertible to `void **`. In other words, currently we can write things like:

```
PyObject *old_exc = _Py_atomic_exchange_ptr(&tstate->async_exc, exc);
```

but if `address` was `void **address`, we'd have to write:

```
PyObject *old_exc = _Py_atomic_exchange_ptr((void **)&tstate->async_exc, exc);
```

Reply:
A large part of the Python C API uses macro to convert arguments to ``PyObject*``. Would it make sense to do the same here?

```
#define _Py_atomic_exchange_ptr(atomic, value) _Py_atomic_exchange_ptr(_Py_CAST(void**, atomic), (value))
```

Reply:
Right, in that case I'm fine with the surprising ``void*`` type.